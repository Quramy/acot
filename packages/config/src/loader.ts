import path from 'path';
import type { ResolvedConfig } from '@acot/types';
import { isDirectorySync } from 'path-type';
import { explorer } from './explorer';
import { debug } from './logging';
import { resolveConfig } from './resolver';

export type LoadConfigOptions = {
  cwd?: string;
};

const searchConfigModule = async (dir: string): Promise<unknown> => {
  debug('search config module: %s', dir);

  const loaded = await explorer.search(dir);
  if (!loaded) {
    throw new ReferenceError('not found config file');
  }

  return loaded.config;
};

const loadConfigModule = async (filepath: string): Promise<unknown> => {
  debug('load config module: %s', filepath);

  try {
    const loaded = await explorer.load(filepath);
    return loaded!.config;
  } catch (e) {
    debug(e);
    throw new ReferenceError('not found config file');
  }
};

export const loadConfig = async (
  from = '.',
  options: LoadConfigOptions = {},
): Promise<ResolvedConfig> => {
  const cwd = options.cwd ?? process.cwd();
  const fileOrDir = path.resolve(cwd, from);
  const isDir = isDirectorySync(fileOrDir);

  debug('path=%s, directory=%s', fileOrDir, isDir);

  const config = await resolveConfig(
    await (isDir ? searchConfigModule(fileOrDir) : loadConfigModule(fileOrDir)),
    { cwd },
  );

  debug('loaded config: %O', config);

  return config;
};
