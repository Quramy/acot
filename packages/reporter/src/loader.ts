import path from 'path';
import { ModuleLoader } from '@acot/module-loader';
import type { ReporterFactory } from '@acot/types';
import { tryResolveModule } from '@acot/utils';

const BUILT_IN_DIR = path.resolve(__dirname, 'reporters');

export class ReporterLoader {
  private _loader: ModuleLoader<ReporterFactory>;

  public constructor(cwd = process.cwd()) {
    this._loader = new ModuleLoader('reporter', {
      from: cwd,
    });
  }

  public load(name: string): ReporterFactory {
    let factory: ReporterFactory | null;

    // built-in
    [factory] = tryResolveModule<ReporterFactory>(
      path.join(BUILT_IN_DIR, name),
    );

    if (factory != null) {
      return factory;
    }

    // package or local files
    factory = this._loader.tryLoad(name);
    if (factory != null) {
      return factory;
    }

    throw new ReferenceError(`The reporter "${name}" was not found.`);
  }
}
