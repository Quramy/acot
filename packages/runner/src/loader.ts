import { ModuleLoader } from '@acot/module-loader';
import type { RunnerFactory } from '@acot/types';

export class RunnerLoader {
  private _loader: ModuleLoader<RunnerFactory>;

  public constructor(cwd = process.cwd()) {
    this._loader = new ModuleLoader('runner', {
      from: cwd,
    });
  }

  public load(name: string): RunnerFactory {
    const factory = this._loader.tryLoad(name);
    if (factory != null) {
      return factory;
    }

    throw new ReferenceError(`The runner "${name}" was not found.`);
  }
}
