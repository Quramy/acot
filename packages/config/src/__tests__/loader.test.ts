import path from 'path';
import { loadConfig } from '../loader';

const FIXTURES_DIR = path.resolve(__dirname, 'fixtures');

describe('loadConfig', () => {
  const opts = {
    cwd: FIXTURES_DIR,
  };

  test('success (search: acot.config.js)', async () => {
    const config = await loadConfig('.', opts);

    expect(config).toMatchSnapshot();
  });

  test('success (load: file.js)', async () => {
    const config = await loadConfig('./file.js', opts);

    expect(config).toMatchSnapshot();
  });

  test('not found config', async () => {
    await expect(
      loadConfig('./__notfound__/', { cwd: __dirname }),
    ).rejects.toThrow(ReferenceError);
  });
});
