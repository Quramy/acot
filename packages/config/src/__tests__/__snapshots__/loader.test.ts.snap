// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadConfig success (load: file.js) 1`] = `
Object {
  "rules": Object {
    "@acot/test/rule1": Array [
      "error",
      null,
    ],
  },
}
`;

exports[`loadConfig success (search: acot.config.js) 1`] = `
Object {
  "rules": Object {
    "@acot/test/rule1": Array [
      "error",
      null,
    ],
    "@acot/test/rule2": Array [
      "warn",
      null,
    ],
    "@acot/test/rule3": Array [
      "off",
      null,
    ],
    "rule1": Array [
      "warn",
      null,
    ],
    "rule2": Array [
      "warn",
      null,
    ],
    "rule3": Array [
      "warn",
      null,
    ],
    "rule4": Array [
      "warn",
      null,
    ],
    "rule5": Array [
      "error",
      null,
    ],
  },
}
`;
